adam-core/src/main/scala/org/bdgenomics/adam/rdd/RegionRDDFunctions.scala

```diff
@@ -0,0 +1,187 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bdgenomics.adam.rdd
+
+import org.apache.spark.SparkContext._
+import org.apache.spark.rdd.RDD
+import org.bdgenomics.adam.models.{ SequenceDictionary, ReferenceRegionWithOrientation, ReferenceRegion }
+import org.bdgenomics.adam.models.ReferenceRegionContext._
+import org.bdgenomics.adam.rich.ReferenceMappingContext.ReferenceRegionReferenceMapping
+import scala.reflect.ClassTag
+import scala.math.max
+
+/**
+ * Functions (joins, filters, groups, counts) that are often invoked on RDD[R] where R is a
+ * type which extends ReferenceRegion (and is often just ReferenceRegion directly).
+ *
+ * @param rdd The argument RDD
+ * @param kt The type of the argument RDD must be explicit
+ * @tparam Region The type of the argument RDD
+ */
+class RegionRDDFunctions[Region <: ReferenceRegion](rdd: RDD[Region])(implicit kt: ClassTag[Region])
+    extends Serializable {
+
+  import RegionRDDFunctions._
+
+  private def regionToReferenceRegion(region: Region): Option[ReferenceRegion] = Option(region)
+
+  private def expander(range: Long)(r: Region): ReferenceRegion =
+    ReferenceRegion(r.referenceName, max(0, r.start - range), r.end + range)
+
+  /**
+   * Performs an 'overlap join' on the input RDD, returning pairs whose first element is a member of the
+   * input RDD and whose second element is a member of the argument RDD which overlaps the first element
+   * of the pair.
+   *
+   * @param that The argument RDD
+   * @param kt2 The type of the values in the argument RDD must be explicit.
+   * @tparam R2 The explicit type of the regions in the argument RDD
+   * @return The RDD of pairs (f, s) where f comes from the input RDD and s comes from 'that', the argument
+   *         RDD.
+   */
+  def joinByOverlap[R2 <: ReferenceRegion](that: RDD[R2])(implicit kt2: ClassTag[R2]): RDD[(Region, R2)] =
+    RegionJoin.partitionAndJoin[Region, R2](rdd, that, r => regionToReferenceRegion(r), (r: R2) => Option(r))
+
+  /**
+   * Performs a 'range join' on the input RDD, returning pairs whose first element is a member of the
+   * input RDD and whose second element is a member of the argument RDD which is within 'range' bases
+   * of the first element of the pair.
+   *
+   * @param that The argument RDD
+   * @param range A long integer; if two regions are within 'range' of each other (along the same chromosome),
+   *              then they are present in the returned RDD.
+   * @param kt2 The type of the values in the argument RDD must be explicit.
+   * @tparam R2 The explicit type of the regions in the argument RDD
+   * @return The RDD of pairs (f, s) where f comes from the input RDD and s comes from 'that', the argument
+   *         RDD.
+   */
+  def joinWithinRange[R2 <: ReferenceRegion](that: RDD[R2], range: Long)(implicit kt2: ClassTag[R2]): RDD[(Region, R2)] =
```

Comment:
Another alternative is to move these fns from here to the corresponding join impls.